1st draft markdown presentation. (#10)
fujitatomoya/presentation: ROS_2_Overview_Introduction/ROS_2_Overview_Introduction.md

---
marp: true
theme: default
header: "__ROS 2 Overview Introduction__"
footer: "[fujitatomoya@github](https://github.com/fujitatomoya) / [tomoyafujita@linkedin](https://www.linkedin.com/in/tomoya-fujita-5bb656b6/)"
paginate: true
---

# ROS 2 Overview Introduction

This is meant to explain ROS 2 basics, design, distribution and features just in overview.
Our centralized mainline documentation site is [ROS 2 documentation](https://docs.ros.org/) if we need to know more details.

<!---
Comment here
--->

---

![bg right:40% 80%](./images/ros2_logo.jpeg)

# Agenda

- Governance
- Distributions
- Architecture / Design
- Basic Concepts
- Samples / Demos
- Remarkable Features

> - 3rd Party Packages

Author: This is not added yet, will be later on.

The ROS Project Management Committee is responsible for the day-to-day operations of the ROS 2 project. The ROS PMC consists of the Project Leader, the ROS PMC Members (who have full voting rights)

See [Current ROS PMC Constituents](https://docs.ros.org/en/rolling/The-ROS2-Project/Governance.html#current-ros-pmc-constituents).

<!---
Comment here
--->

---

## Community Sites

- [ROS 2 Github](https://github.com/ros2)
  - Development, Design, Source code
- [ROS Discourse](https://discourse.ros.org/)
  - Discussion / Announcement / Release
- [Robotics Stack Exchange](https://robotics.stackexchange.com/)
  - Centralized QA platform
- Working Group / Specific Interest Group
  - Under construction since OSRA structure transition.

<!---
Comment here
--->

---

# [Distributions](https://docs.ros.org/en/rolling/Releases.html)

| Distribution Name | Release Date   | End Of Life Date |
| :---------------- | :------------- | :--------------- |
| Kilted Kaiju      | May 2025       | N.A              |
| Jazzy Jalisco     | May 23rd, 2024 | May 2029         |
| Iron Irwini       | May 23rd, 2023 | November 2024    |
| Humble Hawksbill  | May 23rd, 2022 | May 2027         |

> 

Author: remove this line.

<!---
Comment here
--->

---

![bg :100% fit](./images/ros_distros_percentage.png)

<!---
Comment here
--->

---

# Platform & Dependencies

> - C++ Standard C++17 or later

Author: ```suggestion
- C++14 / C++17 Standard
```

<!---
Comment here
--->

---

# Basic Concepts

- [Node](https://docs.ros.org/en/rolling/Concepts/Basic/About-Nodes.html) / [LifecycleNode](https://design.ros2.org/articles/node_lifecycle.html) / [Components](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Composition.html)
- [Executor](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Executors.html)
- [Topics](https://docs.ros.org/en/rolling/Concepts/Basic/About-Topics.html), [Services](https://docs.ros.org/en/rolling/Concepts/Basic/About-Services.html), [Parameters](https://docs.ros.org/en/rolling/Concepts/Basic/About-Parameters.html) and [Actions](https://docs.ros.org/en/rolling/Concepts/Basic/About-Actions.html)
- [Logging](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Logging.html)
- [Launch](https://docs.ros.org/en/rolling/Concepts/Basic/About-Launch.html)

> - [Bagging]()

Author: should be linked with corresponding URL from ros2 documentation.

<!---
Comment here
--->

---

# [Node](https://docs.ros.org/en/rolling/Concepts/Basic/About-Nodes.html)

- A node can be a component
- Single process can be constructed with multi-components
- Component can be loaded during runtime via component container
- Intra-Process Manager

![bg right 100%](https://github.com/ros2/design/raw/gh-pages/img/intraprocess_communication/new_ipc.png)

<!---
Comment here
--->

---

# [Node LifeCycle](https://design.ros2.org/articles/node_lifecycle.html)

> allow roslaunch to ensure that all components have been instantiated correctly before it allows any component to begin executing its behaviour. It will also allow nodes to be restarted or replaced on-line.

Author: ```suggestion
A managed life cycle for nodes allows greater control over the state of ROS system. It allows `roslaunch` to ensure that all components have been instantiated correctly before it allows any component to begin executing its behavior. It will also allow nodes to be restarted or replaced on-line.

```

- Primary State
  - `Unconfigure`, `Inactive`, `Active` and `Finalized`
- Transition State
  - `Configuring`, `ClearningUp`, `ShuttingDown`, `Activating`, `Deactivating` and `ErrorProcessing`

<!---
Comment here
--->

---

![bg 50%](./images/ros2_lifecycle_node.png)

<!---
Comment here
--->

---

# [Components](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Composition.html)

> - Just like ROS1 Nodelet

Author: ```suggestion
- Just like ROS 1 Nodelet
```

- SingleThreadedExecutor: single thread to dispatch task and execute.
- MultiThreadedExecutor: mulith-thread to dispatch task and execute. (via `std::thread::hardware_concurrency`)
- ~~StaticSingleThreadedExecutor~~ (deprecated): statically collect entities for task. that said once spins, it will not collect or update any entities.
- EventExecutor: Uses events queue in the Executor to execute the entities from associated nodes.

<!---
Comment here
--->

---

# [Topics](https://docs.ros.org/en/rolling/Concepts/Basic/About-Topics.html)

![ros2 topic](https://docs.ros.org/en/rolling/_images/Topic-MultiplePublisherandMultipleSubscriber.gif)

<!---
Comment here
--->

---

# [Service](https://docs.ros.org/en/rolling/Concepts/Basic/About-Services.html)

![ros2 service](https://docs.ros.org/en/rolling/_images/Service-MultipleServiceClient.gif)

<!---
Comment here
--->

---

![bg 60%](./images/ros2_service_stack.png)

<!---
Comment here
--->

---

# [Parameter](https://docs.ros.org/en/rolling/Concepts/Basic/About-Parameters.html)

- No parameter server
- Parameters are hosted in Node
- Async / Sync Parameter Client

> - get, set, list, delete and describe methods

Author: ```suggestion
- `get`, `set`, `list`, `delete` and `describe` methods
```

# Composition
ros2 component types
ros2 run rclcpp_components component_container
ros2 component list
ros2 component load /ComponentManager composition composition::Talker
ros2 component load /ComponentManager composition composition::Listener
ros2 component list
--->

---

# Remarkable Features

- [Quality of Service](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html)
- [Physical Domain](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Domain-ID.html) / [Discovery Range Control](https://docs.ros.org/en/rolling/Tutorials/Advanced/Improved-Dynamic-Discovery.html)
- [ROS 2 Security](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Security.html)
- [True Zero Copy / Loaned Message](https://docs.ros.org/en/humble/How-To-Guides/Configure-ZeroCopy-loaned-messages.html)
- Sync / Async Publisher Mode
- Wait for Acknowledgement
- [Content Filtered Topics](https://docs.ros.org/en/rolling/Tutorials/Demos/Content-Filtering-Subscription.html)
- [Network Identifier (ToS, DSCP)](https://design.ros2.org/articles/unique_network_flows.html)
- [Micro-Controller Support (Micro-ROS)](https://micro.ros.org/)

> 

Author: ```suggestion
<!---
Comment here
--->

```

---

> # QoS (Quality of Service)

Author: ```suggestion
# [Quality of Service](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html)
```

- [ROS_DOMAIN_ID](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Domain-ID.html), primary mechanism for having different logical networks share a physical network is known as the Domain ID.

# [Improved Dynamic Discovery](https://docs.ros.org/en/rolling/Tutorials/Advanced/Improved-Dynamic-Discovery.html)

- Discovery protocol initiated in the same subnet By default.
- User can configure the discovery range with environmental variable, such as only localhost, specific peers.

<!---
Comment here
--->

---

# [ROS 2 Security](https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Security.html)

The ability to secure communications among nodes within the ROS 2 computational graph.
Security files enable encryption and authentication, and define policies both for individual nodes and for the overall ROS graph.

see [Setting up security](https://docs.ros.org/en/rolling/Tutorials/Advanced/Security/Introducing-ros2-security.html) for more details.

<!---
Comment here
--->

---

# [Zero Copy Loaned Message](https://docs.ros.org/en/humble/How-To-Guides/Configure-ZeroCopy-loaned-messages.html)

![bg right:30% fit](./images/Fast-DDS_shared_memory_final.png)

> - True Zero Copy / Write Once

Author: ```suggestion
- True Zero Copy / Write Just Once
```

- Configuration
  - Suitable for `PREALLOCATED_MEMORY_MODE` and `PREALLOCATED_WITH_REALLOC_MEMORY_MODE` memory configurations only
  - see [MemoryManagementPolicy](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/rtps/resources/MemoryManagementPolicy.html#_CPPv4N8eprosima8fastrtps4rtps22MemoryManagementPolicy24PREALLOCATED_MEMORY_MODEE)

<!---
Comment here
--->

---

| DDS | Tier | Description |
| :--- | :--- | :--- |
| Fast-DDS | 1 | Zero Copy supported. Even with shared memory transport, DDS feature full supported. |
| Cyclonedds | 1 | DDS with shared memory bypass by iceoryx which requires daemon. When using shared memory, DDS feature cannot be supported such as QoS. |
| RTI Connext DDS | 1 | Commercial only, but most featured DDS implementation |
| Eclipse Iceoryx | 2 | Only shared memory w/o network, daemon process required. |

<!---
Comment here
--->

---

![Fast-DDS_v2.2.0_SHM_Reliable Performance Comparison](./images/Fast-DDS_v2.2.0_SHM_Reliable.png)

<!---
# https://discourse.ros.org/t/fast-dds-v2-2-0-latency-performance/18989
# Intra-Process Communication Performance Improvement by eProsima
--->

---

# Sync / Async Publication

> in ROS aspect, we recommend async send not to block the application threads, so that application can be available for tasks in application perspective. on the other hand, async causes context switches, this could affect the latency.

Author: ```suggestion
in general with ROS 2, we recommend async send not to block the application threads, so that application can be available for tasks in application perspective. on the other hand, async causes context switches, this could affect the latency.
```
